orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+        new_receive.output.extend([new_receive_output_name])
+
+        new_wait.input.extend([new_receive_output_name])
+        new_wait.output.extend([new_wait_output_name])
+
+        for output_node in output_nodes:
+            for i in range(len(output_node.input)):
+                for edgeId in edgeIds:
+                    if output_node.input[i] == edgeId:
+                        output_node.input[i] = new_wait_output_name
+
+    return new_send, new_receive
+
+def find_all_input_nodes(model, node):
+    nodes = []
+    inputs = []
+
+    if node:
+        for inputId in node.input:
+            for node in model.graph.node:
+                for output in node.output:
+                    if output == inputId:
+                        nodes.append(node)
+            for input in model.graph.input:
+                if input.name == inputId:
+                    inputs.append(input)
+    return nodes, inputs
+
+def find_all_output_nodes(model, node):
+    nodes = []
+    outputs = []
+    if node:
+        for outputId in node.output:
+            for node in model.graph.node:
+                for input in node.input:
+                    if input == outputId:
+                        nodes.append(node)
+            for output in model.graph.output:
+                if output.name == outputId:
+                    outputs.append(output)
+    return nodes, outputs
+
+def find_all_output_nodes_by_edge(model, arg):
+    result = []
+    for node in model.graph.node:
+        for input in node.input:
+            if input == arg:
+                result.append(node)
+    return result
+
+# Insert identity nodes to separate same output edge which feeds into different sub-graph.
+def add_identity(model, cuttingEdge, newEdgeIdName):
+    output_nodes = None
+    edgeId = cuttingEdge.edgeId
+    for node in model.graph.node:
+        if len(node.output) >=1 and node.output[0] == edgeId:
+            output_nodes = find_all_output_nodes_by_edge(model, node.output[0])
+            break
+
+    assert output_nodes, "no output node"
+
+    new_identity = model.graph.node.add()
+    new_identity.op_type = 'Identity'
+
+    new_identity.input.extend([edgeId])
+    new_identity.output.extend([newEdgeIdName])
+
+    for i in range(len(output_nodes)):
+        if output_nodes[i].output[0] in cuttingEdge.consumingNodes:
+            for j in range(len(output_nodes[i].input)):
+                if output_nodes[i].input[j] == edgeId:
+                    output_nodes[i].input[j] = newEdgeIdName
+
+    return newEdgeIdName
+
+def find_all_connected_nodes(model, node):
+    nodes0, inputs = find_all_input_nodes(model, node)
+    nodes1, outputs = find_all_output_nodes(model, node)
+
+    connected_nodes = nodes0 + nodes1
+    return connected_nodes, inputs, outputs
+
+def get_node_index(model, node):
+    for i in range(len(model.graph.node)):
+        if model.graph.node[i] == node:
+            return i
+    return None
```

Comment:
if this is used multiple times, maybe better to create a dict from node to index. Besides, this could be simplified to:
return [i for i, n in enumerate(model.graph.node) if n == node][0]